javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
-            .isInstanceOf[BigDecimal]))
-        throw new IllegalFormatConversionException(currentConversionType,
-                                                   arg.getClass())
+        }
+        validateFlags(flags,
+                      conversion,
+                      invalidFlags = InvalidFlagsForOctalAndHex)
 
-      val specialNumberResult = transformFromSpecialNumber()
+      case 'x' | 'X' =>
+        // Hex formatting is not localized
 
-      if (null != specialNumberResult)
-        return specialNumberResult
+        rejectPrecision()
 
-      if ('a' != Character.toLowerCase(currentConversionType)) {
-        formatToken.setPrecision {
-          if (formatToken.isPrecisionSet())
-            formatToken.getPrecision()
-          else
-            FormatToken.DEFAULT_PRECISION
+        val prefix = {
+          if (!flags.altFormat) ""
+          else if (flags.upperCase) "0X"
+          else "0x"
         }
-      }
-
-      val floatUtil = new FloatUtil(
-        result,
-        formatToken,
-        NumberFormat.getInstance(locale).asInstanceOf[DecimalFormat],
-        arg)
-      floatUtil.transform(formatToken, result)
 
-      formatToken.setPrecision(FormatToken.UNSET)
+        def padAndSendWithHexInt(arg: Int): Unit = padAndSendToDest(
+          RootLocaleInfo,
+          flags,
+          width,
+          prefix,
+          applyNumberUpperCase(flags, java.lang.Integer.toHexString(arg))
+        )
 
-      if (getDecimalFormatSymbols().getMinusSign() == result.charAt(0)) {
-        if (formatToken.isFlagSet(FormatToken.FLAG_PARENTHESIS)) {
-          result = wrapParentheses(result)
-          return result.toString()
-        }
-      } else {
-        if (formatToken.isFlagSet(FormatToken.FLAG_SPACE)) {
-          result.insert(0, ' ')
-          startIndex += 1
-        }
-        if (formatToken.isFlagSet(FormatToken.FLAG_ADD)) {
-          result.insert(0, floatUtil.getAddSign())
-          startIndex += 1
+        arg match {
+          case arg: Byte  => padAndSendWithHexInt(arg & 0xFF)
+          case arg: Char  => padAndSendWithHexInt(arg & 0xFFFF)
+          case arg: Short => padAndSendWithHexInt(arg & 0xFFFF)
+          case arg: Int   => padAndSendWithHexInt(arg)
+          case arg: Long =>
+            padAndSendToDest(
+              RootLocaleInfo,
+              flags,
+              width,
+              prefix,
+              applyNumberUpperCase(flags, java.lang.Long.toHexString(arg)))
+          case arg: BigInteger =>
+            formatNumericString(RootLocaleInfo,
+                                flags,
+                                width,
+                                arg.toString(16),
+                                prefix)
+          case _ =>
+            formatNullOrThrowIllegalFormatConversion()
         }
-      }
+        validateFlags(flags,
+                      conversion,
+                      invalidFlags = InvalidFlagsForOctalAndHex)
+      case 'e' | 'E' =>
```

Comment:
Please leave a blank line between all the cases here. Otherwise some of them look more related than others, although that's not true.